### ISSUE TITLE
Android apps should catch & handle JNI crashes as exception

### ISSUE THREAD

**Issue by mhutti1** [kiwix/kiwix-android#62]
_From @kelson42 on September 26, 2016 10:57_

To avoid any wild crashes of the apps. Here is a how to:
http://blog.httrack.com/blog/2013/08/23/catching-posix-signals-on-android/

_Copied from original issue: kiwix/kiwix#330_

**Comment by macgills** [comment]
This requires a kiwix-lib update, the native code must [re-throw exceptions as java ones](https://stackoverflow.com/questions/8415032/catching-exceptions-thrown-from-native-code-running-on-android) for the android app to begin handling them.

**Comment by kelson42** [comment]
@macgills I think:
* Newer version of the libkiwix tends to generate exceptions (general approach)
* The JNI tends to catch them already.

@mhutti1 @macgills Could you confirm? If the libkiwix does not behaves properly from your perspective, please open a ticket and linked it here. But, I believe this ticket can be closed.

**Comment by macgills** [comment]
This ticket cannot be closed. When the native code causes an error we get fatal signals, such errors are present in 2.5.1
![image](https://user-images.githubusercontent.com/3358282/61227965-4250ea00-a71d-11e9-86c3-6dbc6c99b6ab.png)
If Java Exceptions were getting thrown I would get a stacktrace for such an error but what actually happens is I get this
![image](https://user-images.githubusercontent.com/3358282/61228050-67455d00-a71d-11e9-9895-5c2b194500ec.png)
Once it is actually throwing exceptions then on the android side we must catch them.

**Comment by kelson42** [comment]
I think this is a crash of the libzim, so a bug in it. You can not handle that neither in c++ nor in Java IMO. Do you have a reprocase for that?

**Comment by macgills** [comment]
I have absolutely no idea how that crash happened but the library @mhutti1 [linked to](https://github.com/xroche/coffeecatch) claims it can catch SIGSEGV signals and if that was the case I would have a stacktrace and at least some idea of what code execution caused it.

**Comment by kelson42** [comment]
OK, thx for the explanation. Then, yes, if CoffeCatch works, then it should be setup.

**Comment by macgills** [comment]
@mgautierfr I believe this issue should be moved to kiwix-lib, can you confirm this?

**Comment by macgills** [comment]
I am moving this ticket to 3.1, @mgautierfr if you can review the information on this ticket and give your opinion on it's viability we can progress with it

**Comment by mgautierfr** [comment]
Yes it is feasible. At least we will be able to have some trace and solve issues.

Next question will be : "What to do when native code crash ?" If we crash because of a sigsegv, the library is in an undefined state so unusable.

**Comment by macgills** [comment]
Is there much internal state of the library? Can we supply a method to reset the state of the library.
On android we can relaunch the process which is fairly nasty but gives you a guarantee there is no lingering state

**Comment by kelson42** [comment]
@macgills Your last comment would probably benefit of a dedicated ticket on the kiwix-lib repo.

**Comment by kelson42** [comment]
@macgills We can not fix this ticket now. Postponing..

**Comment by kelson42** [comment]
@mgautierfr @macgills I believe it would be good to implement this if this is the path to follow to be able to investigate properly crash reports. Does that sounds doable now?
